crates/bevy_pbr/src/light_probe/prefilter.rs

```diff
@@ -0,0 +1,946 @@
+    pub intensity: f32,
+    pub rotation: Quat,
+    pub affects_lightmapped_mesh_diffuse: bool,
+}
+
+impl Default for FilteredEnvironmentMapLight {
+    fn default() -> Self {
+        FilteredEnvironmentMapLight {
+            environment_map: Handle::default(),
+            intensity: 0.0,
+            rotation: Quat::IDENTITY,
+            affects_lightmapped_mesh_diffuse: true,
+        }
+    }
+}
+
+pub fn extract_prefilter_entities(
+    prefilter_query: Extract<
+        Query<(
+            RenderEntity,
+            &FilteredEnvironmentMapLight,
+            &EnvironmentMapLight,
+        )>,
+    >,
+    mut commands: Commands,
+    render_images: Res<RenderAssets<GpuImage>>,
+) {
+    for (entity, filtered_env_map, env_map_light) in prefilter_query.iter() {
+        let env_map = render_images
+            .get(&filtered_env_map.environment_map)
+            .expect("Environment map not found");
+
+        let diffuse_map = render_images.get(&env_map_light.diffuse_map);
+        let specular_map = render_images.get(&env_map_light.specular_map);
+
+        // continue if the diffuse map is not found
+        if diffuse_map.is_none() || specular_map.is_none() {
+            continue;
+        }
+
+        let diffuse_map = diffuse_map.unwrap();
+        let specular_map = specular_map.unwrap();
+
+        let render_filtered_env_map = RenderEnvironmentMap {
+            environment_map: env_map.clone(),
+            diffuse_map: diffuse_map.clone(),
+            specular_map: specular_map.clone(),
+            intensity: filtered_env_map.intensity,
+            rotation: filtered_env_map.rotation,
+            affects_lightmapped_mesh_diffuse: filtered_env_map.affects_lightmapped_mesh_diffuse,
+        };
+        commands
+            .get_entity(entity)
+            .expect("Entity not synced to render world")
+            .insert(render_filtered_env_map);
+    }
+}
+
+// A render-world specific version of FilteredEnvironmentMapLight that uses CachedTexture
+#[derive(Component, Clone)]
+pub struct RenderEnvironmentMap {
+    pub environment_map: GpuImage,
+    pub diffuse_map: GpuImage,
+    pub specular_map: GpuImage,
+    pub intensity: f32,
+    pub rotation: Quat,
+    pub affects_lightmapped_mesh_diffuse: bool,
+}
+
+#[derive(Component)]
+pub struct PrefilterTextures {
+    pub environment_map: CachedTexture,
+}
+
+/// Prepares textures needed for prefiltering
+pub fn prepare_prefilter_textures(
+    light_probes: Query<Entity, With<RenderEnvironmentMap>>,
+    render_device: Res<RenderDevice>,
+    mut texture_cache: ResMut<TextureCache>,
+    mut commands: Commands,
+) {
+    for entity in &light_probes {
+        // Create environment map with 8 mip levels (512x512 -> 1x1)
+        let environment_map = texture_cache.get(
+            &render_device,
+            TextureDescriptor {
+                label: Some("prefilter_environment_map"),
+                size: Extent3d {
+                    width: 512,
```

Comment:
Do we want to hardcode to 512x512? 256x256 might be cheaper and not worse quality? Or maybe we should let the user configure.